guppylang/compiler/expr_compiler.py

```diff
@@ -452,68 +463,96 @@ def visit_ResultExpr(self, node: ResultExpr) -> Wire:
         return self._pack_returns([], NoneType())
 
     def visit_DesugaredListComp(self, node: DesugaredListComp) -> Wire:
-        from guppylang.compiler.stmt_compiler import StmtCompiler
-
-        compiler = StmtCompiler(self.globals)
-
         # Make up a name for the list under construction and bind it to an empty list
         list_ty = get_type(node)
+        assert isinstance(list_ty, OpaqueType)
         elem_ty = get_element_type(list_ty)
         list_place = Variable(next(tmp_vars), list_ty, node)
-        list_name = with_type(list_ty, with_loc(node, PlaceNode(place=list_place)))
         self.dfg[list_place] = list_new(self.builder, elem_ty.to_hugr(), [])
+        with self._build_generators(node.generators, [list_place]):
+            elt_port = self.visit(node.elt)
+            list_port = self.dfg[list_place]
+            [], [self.dfg[list_place]] = self._build_method_call(
+                list_ty, "append", node, [list_port, elt_port], list_ty.args
+            )
+        return self.dfg[list_place]
+
+    def visit_DesugaredArrayComp(self, node: DesugaredArrayComp) -> Wire:
+        # Allocate an uninitialised array of the desired size and a counter variable
+        array_ty = get_type(node)
+        assert isinstance(array_ty, OpaqueType)
+        array_var = Variable(next(tmp_vars), array_ty, node)
+        count_var = Variable(next(tmp_vars), int_type(), node)
+        hugr_elt_ty = (
+            ht.Option(node.elt_ty.to_hugr())
```

Comment:
It might possibly be worth making this into a method that you call `node.elt_ty.array_element_type()`

Reply:
I think this will be redundant following #630